src/app/qgisapp.cpp

```diff
@@ -10421,7 +10421,38 @@ void QgisApp::mergeSelectedFeatures()
                          vl->dataProvider()->pkAttributeIndexes().contains( vl->fields().fieldOriginIndex( i ) );
 
     if ( isPrimaryKey && !isDefaultValue )
-      mergeFeatureId = val.toLongLong();
+    {
+      QgsFeatureRequest request;
+      request.setFlags( QgsFeatureRequest::Flag::NoGeometry );
+      // Handle multi pks
+      if ( vl->dataProvider()->pkAttributeIndexes().count() > 1 && vl->dataProvider()->pkAttributeIndexes().count() <= attrs.count() )
```

Comment:
@elpaso just noticing that at lines 10417 and 10420 vl->dataProvider() is checked for null. So for consistency this should also be tested here (or the previous tests are useless)

Reply:
ping @elpaso -- any chance we can get this one merged before the final 3.22 release?

Reply:
ping @domi4484 what's the status of this? Have you fixed it in the proper way and backported already or we have to backport this (which was intended as a temporary patch while we are waiting for a proper fix?)

Reply:
> @elpaso just noticing that at lines 10417 and 10420 vl->dataProvider() is checked for null. So for consistency this should also be tested here (or the previous tests are useless)

Done (even if I believe that if data provider is null this method should exit much earlier).

Reply:
@elpaso I have a pr open for that https://github.com/qgis/QGIS/pull/51456.
I should have it ready soon but probably not today. Could we merge this so afterward there will be some more time to test my changes before the next point release?